pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,197 @@
+import random
+
+import torch
+from torch.autograd import Variable
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.util import torch_zeros_like, torch_ones_like
+from pyro.infer import TracePosterior
+from pyro.poutine import Trace
+from tests.common import assert_equal
+
+
+def verlet_integrator(z, r, grad_potential, step_size, num_steps):
+    """
+    Velocity Verlet integrator.
+
+    :param z: trace object containing current values for the sample sites
+    :param r: dictionary of sample site names and corresponding momenta
+    :param grad_potential: function that returns gradient of the potential given z
+        for each sample site
+    :return: (z_next, r_next) having same types as (z, r)
+    """
+    # deep copy the current state - (z, r)
+    z_next = clone_trace(z)
+    r_next = {k: v.clone() for k, v in r.items()}
+    grads = grad_potential(z_next)
+    for _ in range(num_steps):
+        for site_name, node in z_next.iter_stochastic_nodes():
+            # r(n+1/2)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+            # z(n+1)
+            node['value'] = node['value'] + step_size * r_next[site_name]
+            node['value'].retain_grad()
+        grads = grad_potential(z_next)
+        for site_name in r_next:
+            # r(n+1)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+    return z_next, r_next
+
+
+def clone_trace(trace):
+    # clone the tensors so as not to mutate the old state (q, p)
+    # if it is not accepted
+    new_trace = trace.copy()
+    for name, node in trace.iter_stochastic_nodes():
+        node['value'].retain_grad()
```

Comment:
I'm surprised you call `.retain_grad()` before `.clone()`ing, rather than after.

Reply:
There are a couple of funny things going on here, will correct it. I need to find a cleaner way of handling when to `detach`, `retain_grad` and `clone`.